### PR TITLE
Add volID based delete() and resize() if volID is available in pv spec.

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: pkg/volume/glusterfs/glusterfs.go [hunk 1]
```diff
@@ -596,8 +596,16 @@ func (d *glusterfsVolumeDeleter) getGid() (int, bool, error) {
 
 func (d *glusterfsVolumeDeleter) Delete() error {
 	glog.V(2).Infof("delete volume: %s ", d.glusterfsMounter.path)
+	volumeID := ""
 	volumeName := d.glusterfsMounter.path
-	volumeID := dstrings.TrimPrefix(volumeName, volPrefix)
+
```
Comment (dixudx) [inline pkg/volume/glusterfs/glusterfs.go]: Why not having a separate common method?

File: pkg/volume/glusterfs/glusterfs.go [hunk 2]
```diff
@@ -1048,14 +1053,28 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
 	return &cfg, nil
 }
 
+// getVolumeID returns volumeID from the PV or volumename.
+func getVolumeID(pv *v1.PersistentVolume, volumeName string) string {
```
Comment (dixudx) [inline pkg/volume/glusterfs/glusterfs.go]: I'd prefer to return `(string, error)`.

And the error could be `"failed to get volume ID"`.

**Comment by humblec** [comment]
@jsafrane @msau42  Please review :)

File: pkg/volume/glusterfs/glusterfs.go [hunk 3]
```diff
@@ -596,8 +596,16 @@ func (d *glusterfsVolumeDeleter) getGid() (int, bool, error) {
...
 	glog.V(2).Infof("delete volume: %s ", d.glusterfsMounter.path)
+
 	volumeName := d.glusterfsMounter.path
-	volumeID := dstrings.TrimPrefix(volumeName, volPrefix)
+	volumeID, err := getVolumeID(d.spec, volumeName)
+	if err != nil {
+		return fmt.Errorf("failed to get volumeID, err: %v", err)
+	}
+	if volumeID == "" {
+		glog.Errorf("volumeID is nil")
```
Comment (jsafrane) [inline pkg/volume/glusterfs/glusterfs.go]: volumeID is empty, not nil

File: pkg/volume/glusterfs/glusterfs.go [hunk 4]
```diff
@@ -1048,13 +1056,36 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
...
+// getVolumeID returns volumeID from the PV or volumename.
+func getVolumeID(pv *v1.PersistentVolume, volumeName string) (string, error) {
+	volumeID := ""
+
+	// Get volID from pvspec if available, else fill it from volumename.
+	if pv != nil {
+		if pv.Annotations["VolID"] != "" {
+			volumeID = pv.Annotations["VolID"]
+		} else {
+			volumeID = dstrings.TrimPrefix(volumeName, volPrefix)
```
Comment (jsafrane) [inline pkg/volume/glusterfs/glusterfs.go]: What if there is no prefix? Should it fail?

File: pkg/volume/glusterfs/glusterfs.go [hunk 5]
```diff
@@ -1048,13 +1056,36 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
...
+		}
+	} else {
+		return volumeID, fmt.Errorf("Provided PV spec is nil")
```
Comment (jsafrane) [inline pkg/volume/glusterfs/glusterfs.go]: errors start with lowercase

File: pkg/volume/glusterfs/glusterfs.go (hunk 3 above)
Comment (humblec) [reply pkg/volume/glusterfs/glusterfs.go]: done.

File: pkg/volume/glusterfs/glusterfs.go (hunk 4 above)
Comment (humblec) [reply pkg/volume/glusterfs/glusterfs.go]: Volumes were always created with prefix. If its changed with custom volume name it should have an entry in pv spec. What I am trying to say here is  it should fall into any of these conditions.

File: pkg/volume/glusterfs/glusterfs.go (hunk 5 above)
Comment (humblec) [reply pkg/volume/glusterfs/glusterfs.go]: Done.

**Comment by humblec** [comment]
@jsafrane comments are addressed. Thanks !!

File: pkg/volume/glusterfs/glusterfs.go (hunk 4 above)
Comment (jsafrane) [reply pkg/volume/glusterfs/glusterfs.go]: How will the error look like when someone modifies the spec not to include `vol_` or remove annotations? `TrimPrefix` won't do anything, it goes to heketi, heketi says this volume does not exist. Is the error message clear enough for the user to spot the error?

Comment (humblec) [reply pkg/volume/glusterfs/glusterfs.go]: Yes, in that case heketi returns proper error which states the volume does not exist.

**Comment by humblec** [comment]
@jsafrane can you please review this PR ?

**Comment by humblec** [comment]
@jsafrane whenever you take a break from your long review request queue :)

File: pkg/volume/glusterfs/glusterfs.go [hunk 6]
```diff
@@ -1048,13 +1056,36 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
...
 	glog.V(2).Infof("Request to expand volume: %s ", pvSpec.Glusterfs.Path)
 	volumeName := pvSpec.Glusterfs.Path
+	volumeID, err := getVolumeID(spec.PersistentVolume, volumeName)
 
-	// Fetch the volume for expansion.
-	volumeID := dstrings.TrimPrefix(volumeName, volPrefix)
+	if err != nil {
+		return oldSize, fmt.Errorf("failed to get volumeID, err: %v", err)
+	}
+	if volumeID == "" {
```
Comment (jsafrane) [inline pkg/volume/glusterfs/glusterfs.go]: This check is on two places, it should me moved to `getVolumeID`.

Comment (humblec) [reply pkg/volume/glusterfs/glusterfs.go]: Done.

**Comment by humblec** [comment]
@jsafrane Addressed the comment, Thanks!

**Pull request merged**